x/wasm/internal/keeper/keeper_test.go

```diff
@@ -436,38 +436,33 @@ func TestInstantiateWithNonExistingCodeID(t *testing.T) {
 func TestInstantiateWithCallbackToContract(t *testing.T) {
 	ctx, keepers := CreateTestInput(t, false, SupportedFeatures, nil, nil)
 	var (
-		wasmerMock = &selfCallingInstMockWasmer{}
-		err        error
+		excuteCalled bool
+		err          error
 	)
-	keepers.WasmKeeperRef.wasmer = wasmerMock
+	wasmerMock := &MockWasmer{
```

Comment:
I like this constructor over extending `AlwaysFailMockWasmer`. But it does seem a bit odd to inline this.. Fine as now, as it is only used once, but maybe we can make it reusable for other tests. My rough thought:

```go
func selfCallingInstMockWasmer(executeCalled *bool) *MockWasmer {
    return &MockWasmer{

}	CreateFn: func(code cosmwasm.WasmCode) (cosmwasm.CodeID, error) {
			anyCodeID := bytes.Repeat([]byte{0x1}, 32)
			return anyCodeID, nil
		},
		InstantiateFn: func(code cosmwasm.CodeID, env wasmTypes.Env, info wasmTypes.MessageInfo, initMsg []byte, store cosmwasm.KVStore, goapi cosmwasm.GoAPI, querier cosmwasm.Querier, gasMeter cosmwasm.GasMeter, gasLimit uint64) (*wasmTypes.InitResponse, uint64, error) {
			return &wasmTypes.InitResponse{
				Messages: []wasmTypes.CosmosMsg{
					{Wasm: &wasmTypes.WasmMsg{Execute: &wasmTypes.ExecuteMsg{ContractAddr: env.Contract.Address, Msg: []byte(`{}`)}}},
				},
			}, 1, nil
		},
		ExecuteFn: func(code cosmwasm.CodeID, env wasmTypes.Env, info wasmTypes.MessageInfo, executeMsg []byte, store cosmwasm.KVStore, goapi cosmwasm.GoAPI, querier cosmwasm.Querier, gasMeter cosmwasm.GasMeter, gasLimit uint64) (*wasmTypes.HandleResponse, uint64, error) {
			excuteCalled = true
			return &wasmTypes.HandleResponse{}, 1, nil
		},
	}
}
```

And in this test we could do:

```go
executedCalled := false
wasmerMock = selfCallingInstMockWasmer(&executeCalled)
// ...
_, err := keepers.WasmKeeper.Instantiate(ctx, example.CodeID, example.CreatorAddr, nil, nil, "test", nil)
assert.True(t, excuteCalled)
```

Not a blocker and this can always be refactored later. I like the use of MockWasmer.